# Проект 2. Анализ для модели кредитного скоринга

**Цель исследования** - проверить:

Влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

**Задачи исследования** - проверить:

* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**
1. Обзор данных
2. Предобработка данных
3. Анализ данных и решение поставленных задач
4. Итоги исследования

## Обзор данных

Импортируем библиотку pandas

In [1]:
import pandas as pd

Сохраняем файл данных в переменную data

In [2]:
data = pd.read_csv('D:\Documents\Python\ЯПрактикум\Проекты\Проект 2 Кредиты\data.csv')

Выводим на экран первые 10 строк таблицы. 

Сразу заметна проблема с наличием отрицательных значений в столбце "days_employed", а также с дробными значениями в столбце total_income. Вернемся к ним позднее на этапе предобработки данных.

In [3]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получаем общую информацию о таблице методом info()

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов, столбцы названы без нарушений стиля. 

Согласно документации к проекту:

* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

В 2 столбцах обнаруживаются пропуски в данных, данную проблему будет необходимо решить на этапе предобработки данных.

**Выводы**

Предварительно можно утверждать, что данных достаточно для решения поставленных задач. Однако перед проведением анализа необходимо выполнить предобработку данных:

1. Оценить критичность и заполнить пропущенные значения в столбцах days_employed, total_income
2. Найти причину и исправить отрицательные значения в столбце days_employed
3. Округлить значения дохода в столбце total_income, приведя его к целочисленному типу данных
4. Проверить наличие и устранить дубликаты в данных (явные и неявные)
5. Провести необходимые группировки данных для повышения удобства и наглядности дальнейшего анализа 

## Предобработка данных

### Пропуски значений

Посчитаем количество пропущенных значений в таблице:

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В столбцах days_employed и total_income пропущено по 2174 значений, что составляет 10% от общего количества значений в столбцах. Доля пропущенных значений довольно существенна и может повлиять на результаты исследования.

Чтобы понять причины пропусков данных взглянем на несколько строк с пропущенными значениями:

In [6]:
data[data['days_employed'].isna() == True].head() # фильтрация через метод isna и логическую индексацию

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Вначале можно было бы предположить, что значения в столбцах days_employed и total_income могут отсутствовать по причине того, что заемщики пока не успели получить трудовой стаж и находятся в поисках первого места работы. 

Однако при рассмотрении датасета с пропусками видно, что это не так - возраст заемщиков и тип дохода говорит о том, что скорее всего у них был/есть опыт работы и причина пропусков вероятнее всего заключается в том, что заемщики не предоставили данные о своем трудовом стаже и доходе, либо произошла техническая ошибка при обновлении/выгрузки данных.

Пропущенные количественные данные в данном случае лучше всего заполнить медианным значением по столбцу, чтобы исключить влияние выбросов данных (аномально малых и аномально больших значений). Медианное значение лучше отражает усредненное значение по анализируемой совокупности данных без эффекта "средней температуры по больнице".

**Заполнение пропусков в данных**

Заполним пропуски в столбце total_income медианным значением дохода. 

Заполнять пропуски в столбце days_employed, также как исправлять отрицательные значения в этом столбце не будем, так как этот столбец не пригодится для дальнейшего анализа. Кроме того, корректность данных в нем вызывает большие сомнения, так 340 266 дней равняется примерно 932 лет, что не может считаться реальным трудовым стажем.

Рассчитаем медианный доход

In [7]:
median_income = data[data['total_income'].isna() == False]['total_income'].median() # находим медианный доход по значениям в столбце total_income, где нет пропусков
print('Медианный доход:', int(median_income))

Медианный доход: 145017


Заполним пропуски в столбце total_income значением медианного дохода

In [8]:
data['total_income'] = data['total_income'].fillna(median_income) #заполняем пропущенные значения медианным доходом методом fillna
data['days_employed'] = '' #заполняем столбец days_employed пустой строкой
display(data.isna().sum()) # еще раз выводим пропущенные значения по столбцам

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Пропуски в данных устранены.

Заменим вещественные значения в столбце total_income целочисленными.

In [9]:
data['total_income'] = data['total_income'].astype('int') # переводим в целочисленные значения столбец total_income с помощью метода astype
data.info() # проверяем изменение типа данных в столбце total_income

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  object
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(1), int64(5), object(6)
memory usage: 1.9+ MB


Тип данных в столбце total_income изменился на int (целое число).

### Поиск и обработка дубликатов

**Удаляем явные дубликаты**

In [10]:
print('Количество дубликатов:', data.duplicated().sum()) # подсчитаем количество явных дубликатов в данных, найдено 54 дубликата

Количество дубликатов: 54


In [11]:
data = data.drop_duplicates().reset_index(drop = True) # удаляем дубликаты и сбарсываем индексы строк
print('Количество дубликатов:', data.duplicated().sum()) # проверяем, остались ли дубликаты

Количество дубликатов: 0


**Проверяем данные на неявные дубликаты**

Проверку будем осуществлять по каждому столбцу выводя список уникальных значений методом value_counts, таким образом также получится сделать дополнительную проверку на аномальные значения в данных.

* столбец children

In [12]:
data['children'].value_counts() # проверяем первый столбец children методом value_counts, который формирует список уникальных значений и выводит частоту их повторения

 0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Как таковых, дубликатов в столбце children, который отражает количество детей в семье заёмщика быть не может. Но благодаря данной проверке мы выявили аномальные значения в этом столбце: детей точно не может быть -1. 20 детей скорее всего тоже не является реалистичным значением.

Заменим аномальные значениями медианными.

In [13]:
median_children = data['children'].median() # находим медианное значение
print('Медианное количество детей в семье:', int(median_children))

Медианное количество детей в семье: 0


In [14]:
data['children'] = data['children'].replace(-1, 0) # заменяем аномальные значение методом replace
data['children'] = data['children'].replace(20, 0) # заменяем аномальные значение методом replace
print(data['children'].value_counts()) # проверяем итоговый набор уникальных данных

0    14230
1     4809
2     2052
3      330
4       41
5        9
Name: children, dtype: int64


* столбец days_employed : пропускаем как нерелевантный для исследования

* столбец dob_years

In [15]:
data['dob_years'].value_counts()

35    616
40    607
41    606
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    538
48    537
37    536
50    513
43    512
32    509
49    508
28    503
45    497
27    493
52    484
56    484
47    477
54    476
46    473
53    459
57    456
58    456
51    448
59    443
55    443
26    408
60    374
25    357
61    354
62    349
63    269
24    264
64    262
23    253
65    194
22    183
66    182
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Присутствуют записи с аномальным значением возраста клиента 0 лет. Заполним их медианным значением возраста.

In [16]:
median_years = data['dob_years'].median() # находим медианное значение
print('Медианный возраст:', int(median_years))
data['dob_years'] = data['dob_years'].replace(0, median_years) # заменяем аномальные значение методом replace
print(data['dob_years'].value_counts()) # проверяем итоговый набор уникальных данных

Медианный возраст: 42
42    697
35    616
40    607
41    606
34    601
38    597
33    581
39    572
31    559
36    554
44    545
29    544
30    538
48    537
37    536
50    513
43    512
32    509
49    508
28    503
45    497
27    493
52    484
56    484
47    477
54    476
46    473
53    459
58    456
57    456
51    448
59    443
55    443
26    408
60    374
25    357
61    354
62    349
63    269
24    264
64    262
23    253
65    194
22    183
66    182
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


Аномальное значение 0 лет заменено на медианное 42 года

* столбец education

In [17]:
data['education'].value_counts()

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

В столбце education присутствуют неявные дубликаты из-за использования разного регистра. Приведем все записи в этом столбце к нижнему регистру

In [18]:
data['education'] = data['education'].str.lower() # используем метод str.lower для приведения записей в столбце к нижнему регистру
data['education'].value_counts() # контрольный вывод уникальных значений в столбце

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

* столбец family_status

In [19]:
data['family_status'].value_counts()

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

Особых проблем в данном столбце нет, для единообразия можно привести данный столбец также к нижнему регистру

In [20]:
data['family_status'] = data['family_status'].str.lower() # используем метод str.lower для приведения записей в столбце к нижнему регистру
data['family_status'].value_counts() # контрольный вывод уникальных значений в столбце

женат / замужем          12344
гражданский брак          4163
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

столбец gender

In [21]:
data['gender'].value_counts()

F      14189
M       7281
XNA        1
Name: gender, dtype: int64

В данном столбце отсутствуют неявные дубликаты, наличие пропуска/гендера, отличного от "F" / "M" не повлияют на результаты исследования 

* столбец income_type

In [22]:
data['income_type'].value_counts()

сотрудник          11091
компаньон           5080
пенсионер           3837
госслужащий         1457
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

С данными порядок, дубликатов нет

* столбец purpose

In [23]:
data['purpose'].value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Цели получения кредитов разные, однако они поддаются группировке. Создадим отдельные категории по цели получения кредита на этапе категоризации данных.

* cтолбец total_income

как таковых неявных дубликатов в этом столбце не ожидается, но нелишним будет проверить этот столбец на аномальные значения в данных

In [24]:
print('Минимальный доход', data['total_income'].min())
print('Максимальный доход', data['total_income'].max())
print('Средний доход', data['total_income'].mean())
print('Медианный доход', data['total_income'].median())

Минимальный доход 20667
Максимальный доход 2265604
Средний доход 165209.60015835313
Медианный доход 145017.0


Каких-либо аномалий для столбца total_income не выявлено.

Проверка и обработка неявных дубликатов завершена. Были удалены 54 явных дубликата и решена проблема с неявными дубликатами в столбце education. 

Для выявления и удаления явных дубликатов  использовались методы duplicated и drop_duplicates, а также reset_index для сброса индексов строк после удаления дубликатов.

Для выявления неявных дубликатов использовался метод value_counts, который позволяет одновременно вывести уникальные значения столбца и частотность их повторения. Для удаления неявных дубликатов использовались методы приведения записей в столбцах к нижнему регистру str.lower. Также неявные дубликаты в столбце purpose будут обработаны с помощью создания отдельного столбца с категориями цели кредита на этапе категоризации данных. 

Вероятнее всего дубликаты возникли на этапе сбора данных вследствие того, что данные вносились по разному (часть - в верхнем регистре, часть - в нижнем). Также информация о цели кредита, видимо, вносилась клиентами в свободной форме. В будущем необходимо ввести правила заполнения и сбора данных, сделать фиксированные списки для аспектов информации о клиентах, поддающихся категоризации (например, цель кредита - покупка автомобиля, покупка недвижимости, образование).

Также попутно мы проверили столбцы на аномальные значения, которые не были выявлены на ранней стадии подготовки данных.

Параметры датасета после предобработки:

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21471 non-null  int64 
 1   days_employed     21471 non-null  object
 2   dob_years         21471 non-null  int64 
 3   education         21471 non-null  object
 4   education_id      21471 non-null  int64 
 5   family_status     21471 non-null  object
 6   family_status_id  21471 non-null  int64 
 7   gender            21471 non-null  object
 8   income_type       21471 non-null  object
 9   debt              21471 non-null  int64 
 10  total_income      21471 non-null  int32 
 11  purpose           21471 non-null  object
dtypes: int32(1), int64(5), object(6)
memory usage: 1.9+ MB


### Категоризация данных

Выполним категоризацию данных для удобства и наглядности дальнейшего анализа.

**Создание справочников и удаление избыточных столбцов из датасета**

Создадим новые датафреймы - справочники для хранения информации об образовании и семейном положении заемщиков. Это позволит сделать исходную таблицу с данными более лаконичной.

In [26]:
# создадим новые датафреймы путем выбора нужных столбцов из исходного датафрейма и удаления дубликатов, 
#таким образом получим таблицы - справочники

family_status = data[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop = True)
education = data[['education_id', 'education']].drop_duplicates().reset_index(drop = True)

display(family_status, education) # выводим получившиеся датафреймы. 

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

После создания справочников удалим из датасета столбцы education и family status. Также удалим столбец days_employed.

In [27]:
data = data.drop(['education', 'family_status', 'days_employed'], axis = 1) # удаляем столбцы методом drop
data.head() # выводим измененный датасет

children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   

   debt  total_income                     purpose  
0     0        253875               покупка жилья  
1     0        112080     приобретение автомобиля  
2     0        145885               покупка жилья  
3     0        267628  дополнительное образование  
4     0        158616             сыграть свадьбу

**Выделение категорий**

Выделим категории доходов заещиков и напишем функцию для автоматического присвоения категорий:

* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

In [28]:
def total_income_category(total_income):  # создаем функцию для категоризации заемщиков по уровню дохода
    if 0 < total_income <= 30000:
        return 'E'
    if 30000 < total_income <= 50000:
        return 'D'
    if 50000 < total_income <= 200000:
        return 'C'
    if 200000 < total_income < 1000000:
        return 'B'
    if 1000000 < total_income:
        return 'A'

Создадим новый столбец 'total_income_category', который заполним обозначениями категорий дохода

In [29]:
data['total_income_category'] = data['total_income'].apply(total_income_category) # применяем созданную ранее функцию к значениям в столбце total_income с помощью метода apply 
data.head() # выводим обновленный датасет для проверки

children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   

   debt  total_income                     purpose total_income_category  
0     0        253875               покупка жилья                     B  
1     0        112080     приобретение автомобиля                     C  
2     0        145885               покупка жилья                     C  
3     0        267628  дополнительное образование                     B  
4     0        158616             сыграть свадьбу                     C

Создадим функцию для определения категорий назначения кредита. Для определения принадлежности к той или иной категории используем поиск по подстрокам.

* ‘операции с автомобилем’, : 'авто'
* ‘операции с недвижимостью’, : 'недвиж', 'жил'
* ‘проведение свадьбы’, : 'свадьб'
* ‘получение образования’ : 'образован'


In [30]:
def purpose_category(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'недвиж' in purpose or 'жил' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образован' in purpose:
        return 'получение образования'

Создадим новый столбец 'purpose_category', который заполним обозначениями категорий назначения кредита

In [31]:
data['purpose_category'] = data['purpose'].apply(purpose_category) # применяем созданную ранее функцию к значениям в столбце total_income с помощью метода apply 
display(data.head()) # выводим обновленный датасет для проверки
print(data['purpose_category'].value_counts()) # на всякий случай проверим, что ко всем строкам были подобраны категории по ключевым словам (подстрокам)

children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   

   debt  total_income                     purpose total_income_category  \
0     0        253875               покупка жилья                     B   
1     0        112080     приобретение автомобиля                     C   
2     0        145885               покупка жилья                     C   
3     0        267628  дополнительное образование                     B   
4     0        158616             сыграть свадьбу                     C   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64


Предобработка данных завершена, были устранены пропуски в данных, исправлены аномальные значения, проверены и удалены явные и неявные дубликаты, а также выполнена категоризация данных. Можно переходить непосредственно к анализу данных.

## Анализ данных и решение поставленных задач

В данном разделе мы перейдем к непосредственному анализу данных и постараемся ответить на вопросы, обозначенные в начале исследования.

**1. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Сформулируем исходную гипотезу: с увеличением количества детей в семье вероятность возврата кредита в срок снижается.

Напишем функцию, которая будет строить сводную таблицу, которая отражает количество детей и наличие задолженности по кредиту  с помощью метода pivot_table. Аргументом функции будет название столбца датафрейма, который будет источником для строк сводной таблицы.

Основную информацию и значения по столбцам мы проверили на этапе предобработки данных, получения ошибок при выполнении кода не ожидается, поэтому не будем применять обработчик ошибок **try...except...** в данном проекте. 

In [32]:
def def_pivot(pivot_index):
    pivot_def = pd.pivot_table(data, values = ['debt','gender'], index = [pivot_index], aggfunc = {'debt' : 'sum', 'gender' : 'count'}) 
    # строим сводную таблицу с двумя значениями, sum по столбцу debt для подсчета количества заемщиков с задолженностью и count для подсчета общего количества заемщиков в каждой категории
    pivot_def['debt_share'] = (pivot_def['debt'] / pivot_def['gender'] * 100).round(decimals = 1) # добавляем столбец с долей должников от общего количества заемщиков
    pivot_def = pivot_def.rename(columns = {'gender' : 'total'}) # переименовываем столбец с подсчетом общего количества заемщиков в каждой категории
    return pivot_def.sort_values(by = 'debt_share', ascending = False) # выводим итоговую таблицу, отсортированную по убыванию доли заемщиков с задолженностью

In [33]:
display(def_pivot('children'))

debt  total  debt_share
children                         
4            4     41         9.8
2          194   2052         9.5
1          444   4809         9.2
3           27    330         8.2
0         1072  14230         7.5
5            0      9         0.0

В получившейся таблице children отражает количество детей в семье заемщика, debt количество заемщиков, допустивших просрочку по выплатам кредита, non-debt количество заемщиков без просрочек по кредитам и debt_share доля заещиков с просрочкой в общем количестве заемщиков.

Максимальное количество заемщиков с просрочкой не имеет детей, однако делать выводы преждевременно, так как заемщики без детей составляют самое большое количество среди заемщиков. 

Сравним доли заемщиков с просрочкой от общего количества заемщиков. На первом месте с долей 9.8% заемщики с четырьмя детьми, далее располагаются заемщики с двумя, одним и тремя детьми. Доля просрочек по кредитам среди заемщиков без детей составляет 7.5%. Среди полученной статистики из общей закономерности выбивается 9 заемщиков с пятью детьми, не имевших просрочек по кредитам. Взглянем подробнее на данные по таким заемщикам.

In [34]:
display(data[data['children'] == 5])

children  dob_years  education_id  family_status_id gender  \
3978          5         42             1                 1      M   
4395          5         36             1                 0      F   
7861          5         36             1                 0      F   
15795         5         31             1                 0      F   
15889         5         37             1                 0      F   
16183         5         35             1                 1      F   
20403         5         38             3                 0      F   
20787         5         35             1                 0      F   
21104         5         59             1                 0      M   

       income_type  debt  total_income                           purpose  \
3978     сотрудник     0        145017      на покупку своего автомобиля   
4395     компаньон     0        168460          операции с недвижимостью   
7861     сотрудник     0         48772                 операции с жильем   
15795    сотрудник     0         77552  сделка с подержанным автомобилем   
15889    сотрудник     0        256698              покупка недвижимости   
16183  госслужащий     0        126102             на проведение свадьбы   
20403    сотрудник     0        212545      заняться высшим образованием   
20787    компаньон     0        204241                             жилье   
21104    сотрудник     0        269068   операции со своей недвижимостью   

      total_income_category          purpose_category  
3978                      C    операции с автомобилем  
4395                      C  операции с недвижимостью  
7861                      D  операции с недвижимостью  
15795                     C    операции с автомобилем  
15889                     B  операции с недвижимостью  
16183                     C        проведение свадьбы  
20403                     B     получение образования  
20787                     B  операции с недвижимостью  
21104                     B  операции с недвижимостью

Явных проблем и несоответствий с данными строчками не выявлено. у 7 из 9 заемщиков совокупный доход составляет более 100 тыс. руб., что косвенно подтверждает гипотезу о корректности данных. Будем считать, что данные заемщики обладают исключительно 
хорошими навыками финансового планирования и смогли вернуть свои кредиты без просрочек.

**Выводы**

Таким образом, между наличием детей и просрочками по кредиту есть зависимость. Доля должников больше среди заемщиков, имеющих детей и, соответственно, более высокие запланированные и незапланированные расходы. Однако прямой зависимости между количеством детей и долей заемщиков с задолженностью не выявлено, исходная гипотеза подтвердилась лишь частично.

**2. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Исходная гипотеза: разное семейное положение влияет на выплату кредита в срок

Построим соответствующую сводную таблицу для проверки данной гипотезы. За основу возьмем конструкцию из предыдущей задачи, в качестве строк сводной таблицы будет использоваться категория семейного положения.

In [35]:
pivot_family = def_pivot('family_status_id')  # используем уже созданную функцию def_pivot с новым аргументом family_status_id
pivot_family = pivot_family.merge(family_status, how = 'left', on = 'family_status_id') # добавляем семейное положение из ранее созданного справочника family_status
display(pivot_family)

family_status_id  debt  total  debt_share          family_status
0                 4   274   2810         9.8  не женат / не замужем
1                 1   388   4163         9.3       гражданский брак
2                 0   931  12344         7.5        женат / замужем
3                 3    85   1195         7.1              в разводе
4                 2    63    959         6.6         вдовец / вдова

**Выводы**

Самая большая доля задолженности по кредитам наблюдается у заемщиков, не состоящих в браке. При этом меньше всего проблем с возвратом кредитов имеют разведенные заемщики, а также заемщики вдовцы/вдовы. Можно предположить, что неженатые заемщики более легкомысленно подходят к взятию кредита и соблюдению сроков его погашения. В свою очередь, женатые/разведенные заемщики, а тем более заемщики-вдовы/вдовцы более осознанно подходят к такому решению и стараются не допустить возникновение задолженности.

Можно сделать вывод, что возврат кредита в срок зависит от семейного положения, гипотеза подтвердилась.

**3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Исходная гипотеза: Заемщики с более низкими уровнями дохода склонны задерживать выплаты по кредитам, заемщики с более высокими уровнями дохода допускают меньше просрочек по кредитам.

Построим сводную таблицу с детализацией по уровню дохода

In [36]:
display(def_pivot('total_income_category')) # используем уже созданную функцию def_pivot с новым аргументом total_income_category

debt  total  debt_share
total_income_category                         
E                         2     22         9.1
C                      1360  16033         8.5
A                         2     25         8.0
B                       356   5041         7.1
D                        21    350         6.0

**Выводы**

Полученная таблица показывает распределение заемщиков по уровню дохода, а также долю должников в каждой выделенной категории заемщиков. Категория "А" - клиенты с самым высоким уровнем дохода, соответственно, категория "Е" - клиенты с самым низкким уровнем дохода. Наибольшая доля должников в категории "Е", при этом лучше всего возвращают кредиты заемщики из категории "D", которая также характеризуется невысоким уровнем дохода.

Количество заемщиков неравномерно распределено по категориям. Количество заемщиков в категориях "E" и "А" очень низкое по сравнению с количеством заемщиков в других категориях. В случае категории "Е" это может быть вызвано редким одобрением кредита, а в случае категории "А", видимо, у обеспеченных людей реже возникает потребность в кредите. 

При этом гипотеза о том, что заемщики с более высоким уровнем дохода будут реже иметь задолженность по кредитам не подтвердилась, видимо, с ростом дохода растет и сумма кредита.  Зависимости между уровнем дохода и возвратом кредите в срок не наблюдается.

**4. Как разные цели кредита влияют на его возврат в срок?**

Исходная гипотеза: Разные цели кредита влияют на возврат кредита в срок.

Построим сводную таблицу с детализацией по цели кредита

In [37]:
display(def_pivot('purpose_category')) # используем уже созданную функцию def_pivot с новым аргументом purpose_category

debt  total  debt_share
purpose_category                                 
операции с автомобилем     403   4308         9.4
получение образования      370   4014         9.2
проведение свадьбы         186   2335         8.0
операции с недвижимостью   782  10814         7.2

**Выводы**

Распределение заемщиков по категориям цели кредита более равномерное. Самая большая доля просрочек по выплате кредитов среди заемщиков с целью кредита "операции с автомобилем", а самая низкая среди заемщиков с целью "операции с недвижимостью". 
* Высокую долю невозвратов кредита на автомобиль в срок можно объяснить тем, что  покупка автомобиля зачастую сопряжена с дополнительными расходами (покупка резины, установка доп. оборудования и т.п.), не говоря уже о риске повреждении автомобиля в случае аварии и сопряженными с этим финансовыми затратами
* В случае получения образования также существуют риски того, что у заемщика не получится трудоустроиться в намеченный срок, что в итоге приведет к задержке выплаты кредита
* Проведение свадьбы предполагает традиционно предполагает покрытие части расходов за счет подарков гостей, бюджет свадьбы фиксирован и риски невозврата кредита в срок ниже, чем в двух предыдущих вариантах
* Операции с недвижимостью - это ответственный шаг, также часто заемщики оформляют вместе с ипотекой страховку на случай потери трудоспособности. Это может объяснять более низкий процент задержек по выплате кредитов в случае операций с недвижимостью.

Гипотеза о наличии зависимости между целью кредита и возвратом кредита в срок подтвердилась.

## Итоги исследования

Мы проверили гипотезы, сформулированные в начале исследования:

1. Есть ли зависимость между количеством детей и возвратом кредита в срок?

Наличие детей увеличивает вероятность задержки выплаты по кредиту. Но прямой взаимосвязи между количеством детей и задержкой платежа по кредиту не установлено.

2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

Заемщики, не состоящие в браке склонны задерживать выплаты по кредиту по сравнению с женатыми/замужними, либо разведенными заемщиками. Самый низкая доля должников среди вдов/вдовцов. Таким образом между семейным положиением и возвратом кредита в срок существует зависимость.

3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Заемщики с самым низким уровнем дохода хуже остальных возвращают криедиты. Однако в дальнейшем гипотеза не нашла подтверждения, уровень дохода практически не влияет на возврат кредита в срок.

4. Как разные цели кредита влияют на его возврат в срок?

В зависимости от цели кредита меняется доля заемщиков, которые возвращают его в срок. Операции с недвижимостью характеризуется самым низким уровнем просрочек по кредитам, в то время как доля должников по операциям с автомобилем наибольшая. 

Исходная гипотеза о влиянии целей кредита на его возврат в срок подтвердилась.

